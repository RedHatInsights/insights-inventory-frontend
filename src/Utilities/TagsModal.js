import React, { useCallback, useEffect, useState } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchAllTags, toggleTagModal } from '../store/actions';
import { TagModal } from '@redhat-cloud-services/frontend-components/TagModal';
import { cellWidth } from '@patternfly/react-table';
import debounce from 'lodash/debounce';
import flatten from 'lodash/flatten';
import { PAGINATION_DEFAULT } from '../constants';

const TagsModal = ({ filterTagsBy, onToggleModal, onApply, getTags }) => {
  const dispatch = useDispatch();
  const [filterBy, setFilterBy] = useState('');
  const [selected, setSelected] = useState([]);
  const [statePagination, setStatePagination] = useState(PAGINATION_DEFAULT);

  const showTagDialog = useSelector(
    ({ entities, entityDetails }) => (entities || entityDetails)?.showTagDialog
  );

  const pagination = useSelector(({ entities, entityDetails }) => {
    if (entities?.activeSystemTag || entityDetails?.entity) {
      return statePagination;
    }

    return entities?.allTagsPagination || statePagination;
  }, shallowEqual);

  const loaded = useSelector(
    ({ entities, entityDetails }) =>
      entities?.tagModalLoaded || entityDetails?.tagModalLoaded
  );

  const activeSystemTag = useSelector(({ entities, entityDetails }) => {
    if (entityDetails?.entity) {
      return entityDetails?.entity;
    }

    if (entities?.activeSystemTag) {
      return entities.activeSystemTag;
    }
  });

  const tags = useSelector(({ entities, entityDetails }) => {
    const activeTags =
      entities?.activeSystemTag?.tags || entityDetails?.entity?.tags;

    if (activeTags) {
      return activeTags
        ?.filter((tag) =>
          Object.values(tag).some((val) =>
            val?.toLowerCase().includes(filterBy.toLowerCase())
          )
        )
        .slice(
          statePagination?.perPage * (statePagination?.page - 1),
          statePagination?.perPage * statePagination?.page
        );
    }

    return entities?.allTags?.reduce(
      (acc, { tags }) => [...acc, ...flatten(tags.map(({ tag }) => tag))],
      []
    );
  });

  const tagsCount = useSelector(({ entities, entityDetails }) => {
    const activeTags = (
      entities?.activeSystemTag?.tags || entityDetails?.entity?.tags
    )?.filter((tag) =>
      Object.values(tag).some((val) => val?.includes(filterBy))
    );
    return activeTags ? activeTags.length : entities?.allTagsTotal;
  });

  useEffect(() => {
    setFilterBy(filterTagsBy);
  }, [filterTagsBy]);

  const fetchTags = useCallback(
    (pagination, filterBy) => {
      if (!activeSystemTag) {
        dispatch(fetchAllTags(filterBy, pagination, getTags));
      } else {
        setStatePagination(() => pagination);
      }
    },
    [activeSystemTag, dispatch, getTags]
  );

  const debouncedFetch = useCallback(
    () => debounce(fetchTags, 800),
    [fetchTags]
  );

  return (
    <TagModal
      className="ins-c-inventory__tags-modal"
      tableProps={{
        canSelectAll: false,
      }}
      {...(loaded && {
        loaded,
        pagination: {
          ...pagination,
          count: tagsCount,
        },
        rows:
          tags?.map(({ key, value, namespace }) => ({
            id: `${namespace}/${key}=${value}`,
            selected: selected.find(
              ({ id }) => id === `${namespace}/${key}=${value}`
            ),
            cells: [key, value, namespace],
          })) || [],
      })}
      loaded={loaded}
      isOpen={showTagDialog}
      toggleModal={() => {
        setSelected([]);
        setFilterBy('');
        onToggleModal();
        setStatePagination(PAGINATION_DEFAULT);
        dispatch(toggleTagModal(false));
      }}
      filters={[
        {
          label: 'Tags filter',
          placeholder: 'Filter tags',
          value: 'tags-filter',
          type: 'text',
          filterValues: {
            value: filterBy,
            onChange: (_e, value) => {
              debouncedFetch(pagination, value);
              setFilterBy(value);
            },
          },
        },
      ]}
      onUpdateData={(pagination) => fetchTags(pagination, filterBy)}
      columns={[
        { title: 'Name' },
        { title: 'Value', transforms: [cellWidth(30)] },
        { title: 'Tag source', transforms: [cellWidth(30)] },
      ]}
      {...(!activeSystemTag && {
        onSelect: (selected) => setSelected(selected),
        selected,
        onApply: () => onApply && onApply(selected),
      })}
      bulkSelect={{ id: 'bulk-select-tags' }}
      title={
        activeSystemTag
          ? `${
              activeSystemTag.display_name || activeSystemTag.name
            } (${tagsCount})`
          : `All tags in inventory (${tagsCount})`
      }
    />
  );
};

TagsModal.propTypes = {
  onApply: PropTypes.func,
  onToggleModal: PropTypes.func,
  filterTagsBy: PropTypes.string,
  customFilters: PropTypes.shape({
    tags: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.string),
    ]),
  }),
  getTags: PropTypes.func,
};

TagsModal.defaultProps = {
  filterTagsBy: '',
  onToggleModal: () => undefined,
};

export default TagsModal;
