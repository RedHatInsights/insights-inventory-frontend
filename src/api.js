import { INVENTORY_API_BASE } from './config';

const addMockApp = (results) => results.map(row => ({
    id: row.id,
    health: {
        vulnerabilities: { title: 5, redirect: 'cost_management' },
        configuration: { title: 10, redirect: 'configuration_assessment' },
        compliance: { title: '74%', redirect: 'compliance' },
        cost: { title: '23K', redirect: 'cost_management' }
    }
}));

export function getEntities () {
    return fetch(INVENTORY_API_BASE).then(r => {
        if (r.ok) {
            return r.json().then(data => ({
                results: addMockApp(data.results)
            }));
        }

        throw new Error(`Unexpected response code ${r.status}`);
    });
}

export function getEntity (id) {
    return fetch(`${INVENTORY_API_BASE}/${id}`).then(r => {
        if (r.ok) {
            return r.json();
        }

        throw new Error(`Unexpected response code ${r.status}`);
    });
}
