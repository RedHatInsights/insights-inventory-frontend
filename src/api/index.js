export { default } from './api';
export * from './api';
export * from './systemIssues';
