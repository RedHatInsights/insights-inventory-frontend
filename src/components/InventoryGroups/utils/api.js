import { instance } from '@redhat-cloud-services/frontend-components-utilities/interceptors/interceptors';
import { INVENTORY_API_BASE } from '../../../api';
import { TABLE_DEFAULT_PAGINATION } from '../../../constants';
import PropTypes from 'prop-types';
import fixtureGroups from '../../../../cypress/fixtures/groups.json';
import fixtureGroupsDetails from '../../../../cypress/fixtures/groups/Ba8B79ab5adC8E41e255D5f8aDb8f1F3.json';

// eslint-disable-next-line camelcase
export const getGroups = (search = {}, pagination = { page: 1, per_page: TABLE_DEFAULT_PAGINATION }) => {
    if (window.Cypress) {
        const parameters = new URLSearchParams({
            ...search,
            ...pagination
        }).toString();

        return instance.get(`${INVENTORY_API_BASE}/groups?${parameters}`);
    }

    // FIXME: remove mock data when API is implemented
    return Promise.resolve(fixtureGroups);
};

export const createGroup = (payload) => {
    return instance.post(`${INVENTORY_API_BASE}/groups`, {
        name: payload.name,
        // eslint-disable-next-line camelcase
        host_ids: []
    });
};

export const validateGroupName = (name) => {
    if (window.Cypress) {
        return instance.get(`${INVENTORY_API_BASE}/groups`)
        .then((resp) => resp?.results.some((group) => group.name === name));
    }

    // FIXME: remove mock data when API is implemented
    return Promise.resolve(fixtureGroups).then((resp) => resp?.results.some((group) => group.name === name));
};

export const getGroupDetail = (groupId) => {
    if (window.Cypress) {
        return instance.get(`${INVENTORY_API_BASE}/groups/${groupId}`);
    }

    // FIXME: remove mock data when API is implemented
    return Promise.resolve(fixtureGroupsDetails);
};

export const updateGroupById = (id, payload) => {
    return instance.patch(`${INVENTORY_API_BASE}/groups/${id}`, {
        name: payload.name
    });
};

export const deleteGroupsById = (ids = []) => {
    return instance.delete(`${INVENTORY_API_BASE}/groups/${ids.join(',')}`);

};

getGroups.propTypes = {
    search: PropTypes.shape({
    // eslint-disable-next-line camelcase
        hostname_or_id: PropTypes.string
    }),
    pagination: PropTypes.shape({
        perPage: PropTypes.number,
        page: PropTypes.number
    })
};

getGroupDetail.propTypes = {
    groupId: PropTypes.string.isRequired
};
