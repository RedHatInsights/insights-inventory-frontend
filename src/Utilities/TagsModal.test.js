/* eslint-disable no-import-assign */
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import debounce from 'lodash/debounce';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { createPromise as promiseMiddleware } from 'redux-promise-middleware';
import * as api from '../api/api';
import TagsModal from './TagsModal';

jest.mock('lodash/debounce');
describe('TagsModal', () => {
  let initialState;
  let mockStore;
  beforeEach(() => {
    debounce.mockImplementation(jest.requireActual('lodash/debounce'));
    mockStore = configureStore([promiseMiddleware()]);
    initialState = {
      entities: {
        showTagDialog: true,
        tagModalLoaded: true,
      },
    };
  });
  describe('DOM', () => {
    it('should render loading state correctly', () => {
      const store = mockStore({
        entities: {
          showTagDialog: true,
          tagModalLoaded: false,
          activeSystemTag: {
            tagsLoaded: false,
          },
        },
      });
      const view = render(
        <Provider store={store}>
          <TagsModal />
        </Provider>
      );

      // TODO: improve skeleton accessibility
      // baseElement is used since the PF modal is rendered outside parent div
      expect(view.baseElement).toMatchSnapshot();
      expect(
        screen.getByRole('columnheader', {
          name: /name/i,
        })
      ).toBeVisible();
      expect(
        screen.getByRole('columnheader', {
          name: /value/i,
        })
      ).toBeVisible();
      expect(
        screen.getByRole('columnheader', {
          name: /tag source/i,
        })
      ).toBeVisible();
    });

    it('should render activeSystemTag', () => {
      const store = mockStore({
        entities: {
          ...initialState.entities,
          activeSystemTag: {
            tags: [
              {
                key: 'some',
                value: 'test',
                namespace: 'something',
              },
            ],
            tagsLoaded: true,
            tagsCount: 50,
            page: 1,
            perPage: 10,
          },
        },
      });
      render(
        <Provider store={store}>
          <TagsModal />
        </Provider>
      );

      expect(
        screen.getByRole('columnheader', {
          name: /name/i,
        })
      ).toBeVisible();
      expect(
        screen.getByRole('columnheader', {
          name: /value/i,
        })
      ).toBeVisible();
      expect(
        screen.getByRole('columnheader', {
          name: /tag source/i,
        })
      ).toBeVisible();

      expect(screen.getAllByRole('cell')).toHaveLength(3);
      screen
        .getAllByRole('cell')
        .forEach((cell, index) =>
          expect(cell).toHaveTextContent(['some', 'test', 'something'][index])
        );
    });

    it('should render all tags', () => {
      const store = mockStore({
        entities: {
          ...initialState.entities,
          allTagsLoaded: true,
          allTagsTotal: 50,
          allTagsPagination: {
            page: 1,
            perPage: 10,
          },
          allTags: [
            {
              tags: [
                {
                  tag: {
                    key: 'some',
                    value: 'test',
                    namespace: 'something',
                  },
                },
              ],
            },
          ],
        },
      });
      render(
        <Provider store={store}>
          <TagsModal store={store} />
        </Provider>
      );

      expect(screen.getByText(/all tags in inventory \(50\)/i)).toBeVisible();
      expect(
        screen.getByRole('button', {
          name: /apply tags/i,
        })
      ).toBeDisabled();
      expect(
        screen.getByRole('button', {
          name: /cancel/i,
        })
      ).toBeEnabled();
    });
  });

  describe('API', () => {
    beforeEach(() => {
      api.getTags = jest.fn().mockImplementation(() => Promise.resolve());
      api.getAllTags = jest.fn().mockImplementation(() => Promise.resolve());
    });

    it('should call onApply select correct tag', async () => {
      const onApply = jest.fn();
      const store = mockStore({
        entities: {
          ...initialState.entities,
          allTagsLoaded: true,
          allTagsTotal: 50,
          allTagsPagination: {
            page: 1,
            perPage: 10,
          },
          allTags: [
            {
              tags: [
                {
                  tag: {
                    key: 'some',
                    value: 'test',
                    namespace: 'something',
                  },
                },
              ],
            },
          ],
        },
      });
      render(
        <Provider store={store}>
          <TagsModal onApply={onApply} />
        </Provider>
      );

      await userEvent.click(
        screen.getByRole('checkbox', {
          name: /select row 0/i,
        })
      );
      await userEvent.click(
        screen.getByRole('button', {
          name: /apply tags/i,
        })
      );
      expect(onApply).toHaveBeenCalledTimes(1);
    });

    it('should toggle modal', async () => {
      const store = mockStore({
        entities: {
          ...initialState.entities,
          allTagsLoaded: true,
          allTagsTotal: 50,
          allTagsPagination: {
            page: 1,
            perPage: 10,
          },
          allTags: [
            {
              tags: [
                {
                  tag: {
                    key: 'some',
                    value: 'test',
                    namespace: 'something',
                  },
                },
              ],
            },
          ],
        },
      });
      render(
        <Provider store={store}>
          <TagsModal />
        </Provider>
      );

      await userEvent.click(
        screen.getByRole('button', {
          name: /close/i,
        })
      );
      const actions = store.getActions();
      expect(actions[0]).toMatchObject({
        payload: { isOpen: false },
        type: 'TOGGLE_TAG_MODAL',
      });
    });

    it('should fetch additional tags when all tags shown', async () => {
      const store = mockStore({
        entities: {
          ...initialState.entities,
          allTagsLoaded: true,
          allTagsTotal: 50,
          allTagsPagination: {
            page: 1,
            perPage: 10,
          },
          allTags: [
            {
              tags: [
                {
                  tag: {
                    key: 'some',
                    value: 'test',
                    namespace: 'something',
                  },
                },
              ],
            },
          ],
        },
      });
      render(
        <Provider store={store}>
          <TagsModal />
        </Provider>
      );

      expect(
        screen.getAllByRole('button', { name: /go to next page/i })
      ).toHaveLength(2);
      await userEvent.click(
        screen.getAllByRole('button', { name: /go to next page/i })[0]
      );
      const actions = store.getActions();
      expect(actions[0]).toMatchObject({ type: 'ALL_TAGS_PENDING' });
    });
  });
});
