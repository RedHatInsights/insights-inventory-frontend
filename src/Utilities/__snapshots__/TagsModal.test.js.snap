// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsModal DOM should render loading state correctly 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-2"
          aria-labelledby="pf-modal-part-1"
          aria-modal="true"
          class="pf-c-modal-box ins-c-tag-modal ins-c-inventory__tags-modal pf-m-md"
          data-ouia-component-id="OUIA-Generated-Modal-medium-1"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-0"
          role="dialog"
          style="width: auto;"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-medium-1-ModalBoxCloseButton"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          <header
            class="pf-c-modal-box__header"
          >
            <h1
              class="pf-c-modal-box__title"
              id="pf-modal-part-1"
            >
              
              <span
                class="pf-c-modal-box__title-text"
              >
                All tags in inventory (undefined)
              </span>
            </h1>
          </header>
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-2"
          >
            <div
              class="pf-c-toolbar  ins-c-primary-toolbar"
              data-ouia-component-id="PrimaryToolbar"
              data-ouia-component-type="PF4/Toolbar"
              data-ouia-safe="true"
              id="ins-primary-data-toolbar"
            >
              <div
                class="pf-c-toolbar__content"
              >
                <div
                  class="pf-c-toolbar__content-section"
                >
                  <div
                    class="pf-c-toolbar__group pf-m-filter-group ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
                  >
                    <div
                      class="pf-c-toolbar__item"
                    >
                      <div
                        class="pf-c-dropdown ins-c-bulk-select"
                        data-ouia-component-id="BulkSelect"
                        data-ouia-component-type="PF4/Dropdown"
                        data-ouia-safe="true"
                      >
                        <div
                          class="pf-c-dropdown__toggle pf-m-split-button"
                        >
                          <label
                            class="pf-c-dropdown__toggle-check"
                            for="bulk-select-tags-toggle-checkbox"
                          >
                            <input
                              aria-invalid="false"
                              aria-label="Select all"
                              data-ouia-component-id="BulkSelect"
                              data-ouia-component-type="PF4/DropdownToggleCheckbox"
                              data-ouia-safe="true"
                              id="bulk-select-tags-toggle-checkbox"
                              type="checkbox"
                            />
                            
                          </label>
                          <button
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-label="Select"
                            class="pf-c-dropdown__toggle-button"
                            data-ouia-component-id="BulkSelect"
                            data-ouia-component-type="PF4/DropdownToggle"
                            data-ouia-safe="true"
                            id="pf-dropdown-toggle-id-0"
                            type="button"
                          >
                            <span
                              class=""
                            >
                              <svg
                                aria-hidden="true"
                                fill="currentColor"
                                height="1em"
                                role="img"
                                style="vertical-align: -0.125em;"
                                viewBox="0 0 320 512"
                                width="1em"
                              >
                                <path
                                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="pf-c-toolbar__item ins-c-primary-toolbar__filter"
                    >
                      <div
                        class="pf-l-split ins-c-conditional-filter"
                      >
                        <div
                          class="pf-l-split__item pf-m-fill"
                        >
                          <input
                            aria-invalid="false"
                            aria-label="text input"
                            class="pf-c-form-control ins-c-conditional-filter "
                            data-ouia-component-id="ConditionalFilter"
                            data-ouia-component-type="PF4/TextInput"
                            data-ouia-safe="true"
                            placeholder="Filter tags"
                            type="text"
                            value=""
                            widget-type="InsightsInput"
                          />
                          <svg
                            aria-hidden="true"
                            class="ins-c-search-icon"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            style="vertical-align: -0.125em;"
                            viewBox="0 0 512 512"
                            width="1em"
                          >
                            <path
                              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="pf-c-toolbar__item ins-c-primary-toolbar__pagination"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__lg"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="pf-c-toolbar__expandable-content"
                  id="ins-primary-data-toolbar-expandable-content-0"
                >
                  <div
                    class="pf-c-toolbar__group"
                  />
                </div>
              </div>
              <div
                class="pf-c-toolbar__content pf-m-hidden"
                hidden=""
              >
                <div
                  class="pf-c-toolbar__group"
                />
              </div>
            </div>
            <table
              aria-label="Loading"
              class="pf-c-table pf-m-grid-md"
              data-ouia-component-id="OUIA-Generated-Table-2"
              data-ouia-component-type="PF4/Table"
              data-ouia-safe="true"
              role="grid"
            >
              <thead
                class=""
              >
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-1"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <th
                    class=""
                    data-key="0"
                    data-label="Name"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    class="pf-m-width-30"
                    data-key="1"
                    data-label="Value"
                    scope="col"
                  >
                    Value
                  </th>
                  <th
                    class="pf-m-width-30"
                    data-key="2"
                    data-label="Tag source"
                    scope="col"
                  >
                    Tag source
                  </th>
                </tr>
              </thead>
              <tbody
                class=""
                role="rowgroup"
              >
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-2"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-3"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-4"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-5"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-6"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-7"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-8"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-9"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-10"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  class=""
                  data-ouia-component-id="OUIA-Generated-TableRow-11"
                  data-ouia-component-type="PF4/TableRow"
                  data-ouia-safe="true"
                >
                  <td
                    class=""
                    data-key="0"
                    data-label="Name"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="1"
                    data-label="Value"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                  <td
                    class=""
                    data-key="2"
                    data-label="Tag source"
                  >
                    <div
                      class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
                    >
                      <span
                        class="pf-u-screen-reader"
                      />
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <footer
            class="pf-c-modal-box__footer"
          >
            <button
              aria-disabled="true"
              class="pf-c-button pf-m-primary pf-m-disabled"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              disabled=""
              type="button"
            >
              Apply 
              tags
            </button>
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Cancel
            </button>
          </footer>
        </div>
      </div>
    </div>
  </div>
</body>
`;
